app/src/main/java/com/example/setgame/GameActivity.kt

```diff
@@ -0,0 +1,263 @@
+            }
+            else if (gameTemp.previousChosenCards.isNotEmpty() && game.previousChosenCards.isNotEmpty()) {
+                if (gameTemp.previousChosenCards[0] != game.previousChosenCards[0]) {
+                    game = gameTemp
+                    changeCards()
+                }
+            }
+            else if (gameTemp.previousChosenCards.isNotEmpty() && game.previousChosenCards.isEmpty()) {
+
+                    game = gameTemp
+                    changeCards()
+            }
+
+            else if(gameTemp.cards.size == 0){
+                getScores()
+            }
+            if (gameTemp.extraRows != game.extraRows || game.extraRows != extraRows){
+                game = gameTemp
+                addExtraRow()
+            }
+        }
+        override fun onCancelled(databaseError: DatabaseError) {
+            Log.w(TAG, "loadPost:onCancelled", databaseError.toException())
+        }
+    }
+
+    fun createField(){
+        this.setContentView(R.layout.game_activity)
+        findViewById<TextView>(R.id.textView2).setText(gameId.toString())
+        scoresLayout = findViewById<LinearLayout>(R.id.scoresTable)
+        scoresText = TextView(this)
+        var temp = ""
+        for(v in game.scores.keys.toList()){
+            temp = temp + v.toString() + " -> "+ game.scores[v].toString() + "\n"
+        }
+        scoresText.setText(temp)
+        scoresLayout.addView(scoresText)
+        val layout = findViewById<TableLayout>(R.id.tableLayout)
+//        layout.layoutParams = TableLayout.LayoutParams(TableLayout.LayoutParams.MATCH_PARENT, TableLayout.LayoutParams.MATCH_PARENT)
+//        layout.setPadding(3, 3, 3, 3)
+        cardsArray = List(0, {0})
+        for (i in 0..3) {
+            val tr = TableRow(this)
+            tr.setLayoutParams(
+                    TableRow.LayoutParams(
+                            TableRow.LayoutParams.FILL_PARENT,
+                            TableRow.LayoutParams.WRAP_CONTENT,
+                            1.0f
+                    )
+            )
+            val metrics = DisplayMetrics()
+            windowManager.defaultDisplay.getMetrics(metrics)
+
+            val width = (metrics.widthPixels - 30) / 3
+            val height = (width / 1.81).toInt();
+
+            for (j in 0..2) {
+                val temp = ImageButton(this)
+                val cardCode = game.popCard()
+                var bmp = MediaStore.Images.Media.getBitmap(this.getContentResolver(), Uri.parse("android.resource://com.example.setgame/drawable/c$cardCode"));
+                bmp = Bitmap.createScaledBitmap(bmp, width, height.toInt(), true);
+                temp.id = cardCode!!;
+                temp.setBackgroundColor(Color.WHITE);
+                temp.setImageBitmap(bmp)
+                temp.setPadding(2, 2, 2, 2)
+                temp.setOnClickListener(this)
+                tr.addView(temp, j)
+            }
+            layout.addView(tr, i)
+        }
```

Comment:
The main thing: instead of manual management of such things use RecyclerView (with GridLayoutManager), Adapter and ViewHolders.